mock_drivers/rosbag_mock_drivers/src/USFControllerDriver.cpp

```diff
@@ -0,0 +1,249 @@
+{
+  return 20;  // 20 Hz as default spin rate to match expected controller data rate
+}
+
+//USF controller message structure
+struct CommandData {
+  double linear_x;
+  double angular_z;
+
+  void reset();
+};
+
+void CommandData::reset()
+{
+  linear_x      = 0;
+  angular_z     = 0;
+}
+
+static CommandData command_data;
+
+static void vehicleCmdCallback(const autoware_msgs::VehicleCmd& msg)
+{
+  command_data.linear_x = msg.twist_cmd.twist.linear.x;
+  command_data.angular_z = msg.twist_cmd.twist.angular.z;
+}
+
+static void *sendCommand(void *arg)
+{
+  int *client_sockp = static_cast<int*>(arg);
+  int client_sock = *client_sockp;
+  delete client_sockp;
+
+  std::ostringstream oss;
+  oss << command_data.linear_x << ",";
+  oss << command_data.angular_z;
+
+  std::string cmd(oss.str());
+  ssize_t n = write(client_sock, cmd.c_str(), cmd.size());
+  if(n < 0){
+    std::perror("write");
+    return nullptr;
+  }
+
+  if(close(client_sock) == -1){
+    std::perror("close");
+    return nullptr;
+  }
+
+  std::cout << "cmd: " << cmd << ", size: " << cmd.size() << std::endl;
+  return nullptr;
+}
+
+static void* receiverCaller(void *unused)
+{
+  constexpr int listen_port = 10001;
+
+  int sock = socket(AF_INET, SOCK_STREAM, 0);
+  if(sock == -1){
+    std::perror("socket");
+    return nullptr;
+  }
+
+  sockaddr_in addr;
+  sockaddr_in client;
+  socklen_t len = sizeof(client);
+
+  std::memset(&addr, 0, sizeof(sockaddr_in));
+  addr.sin_family = PF_INET;
+  addr.sin_port = htons(listen_port);
+  addr.sin_addr.s_addr = INADDR_ANY;
+
+  int ret = bind(sock, (struct sockaddr *)&addr, sizeof(addr));
+  if(ret == -1){
+    std::perror("bind");
+    goto error;
+  }
+
+  ret = listen(sock, 20);
+  if(ret == -1){
+    std::perror("listen");
+    goto error;
+  }
+  while(true){
+    std::cout << "Waiting access..." << std::endl;
+
+    int *client_sock = new int();
+    *client_sock = accept(sock, reinterpret_cast<sockaddr*>(&client), &len);
+    if(*client_sock == -1){
+      std::perror("accept");
+      break;
+    }
+
+    std::cout << "get connect." << std::endl;
+
+    pthread_t th;
+    if(pthread_create(&th, nullptr, sendCommand, static_cast<void*>(client_sock)) != 0){
+      std::perror("pthread_create");
+      break;
+    }
+
+    if(pthread_detach(th) != 0){
+      std::perror("pthread_detach");
+      break;
+    }
+  }
+
+  error:
+  close(sock);
+  return nullptr;
+}
+
+int MockControllerDriver::onRun()
+{
+  // driver publisher, subscriber, and service
+  mock_driver_node_.addPub(robot_status_ptr_);
+  //mock_driver_node_.addSub(vehicle_cmd_ptr_); //vehicle command topic will be subscribed later.
+  mock_driver_node_.addSrv(enable_robotic_ptr_);
+  MockControllerDriver::robot_enabled_ = true; //always set robot_enabled to be true
+  
+  
+  //USF customized DBW control codes
+  int argc;
+  char **argv;
+
+  ros::init(argc, argv, "vehicle_sender"); //define a node to receive control command and send it to DBW using TCP/IP
```

Comment:
I think ultimately we'd want it to be it's own standalone node right? The mock_drivers aspect here is just a template since they wanted a node that had a lot of the topics already setup. Ideally this'd just be another driver in our system similar to the ssc_interface_wrapper.

Reply:
Has been removed.